### PR TITLE
Feat/version2.0 feature merges

### PR DESCRIPTION
**Pull request by lfbrehm**
## Summary
This PR contains delete, relations and archive requests.
## Delete
- Resources are updated with `object_status = DELETED`
## Relations
- The API request `ModifyRelations` is now implemented
## Archive/Snapshot
- Projects and all resources belonging to the object are updated to `dynamic = false`
- Collections/Datasets are cloned with a persistent copy

### CODE REVIEW

File: src/middlelayer/snapshot_request_types.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    }
+    async fn get_all_relation_ids(
+        project_id: DieselUlid,
+        client: &Client,
+    ) -> Result<Vec<DieselUlid>> {
+        let mut relations: Vec<DieselUlid> = Vec::new();
+        let (mut ids, mut init_in) = InternalRelation::get_all_by_id(project_id, client).await?;
+        ids.append(&mut init_in);
+        for id in ids.iter().map(|i| i.id) {
+            let (outs, ins) = InternalRelation::get_all_by_id(id, client).await?;
```
Comment (St4NNi) [inline src/middlelayer/snapshot_request_types.rs:217]: get_all in for loop ?

**Review by St4NNi — CHANGES_REQUESTED** [review]
Good work, one thing i really do not like is iterating over DB requests. We should add get / update requests that take a list and process them all at once.

File: src/grpc/projects.rs [hunk 2]
```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
...
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_name(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```
Comment (das-Abroxas) [inline src/grpc/projects.rs:151]: Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_name(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```

File: src/grpc/projects.rs [hunk 3]
```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
...
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_description(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```
Comment (das-Abroxas) [inline src/grpc/projects.rs:184]: Copy & paste only partially adapted.

```suggestion
        let request = DescriptionUpdate::Project(request.into_inner());
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_description(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```

File: src/grpc/projects.rs [hunk 4]
```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
...
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_keyvals(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```
Comment (das-Abroxas) [inline src/grpc/projects.rs:218]: Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_keyvals(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```

File: src/grpc/projects.rs [hunk 5]
```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
...
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_dataclass(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```
Comment (das-Abroxas) [inline src/grpc/projects.rs:252]: Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_dataclass(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```

File: src/grpc/projects.rs [hunk 6]
```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
...
+            self.database_handler.snapshot(request).await,
+            "Internal database error."
+        );
+        for resource in &resources {
+            self.cache
+                .update_object(&resource.object.id, resource.clone());
+        }
+        let project: generic_resource::Resource =
+            // First entry is always the archived project 
+            tonic_internal!(resources[0].clone().try_into(), "Dataset conversion error");
```
Comment (das-Abroxas) [inline src/grpc/projects.rs:289]: Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::ADMIN, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        // This contains all the resources the snapshot includes below the project
        let resources = tonic_internal!(
            self.database_handler.snapshot(request).await,
            "Internal database error."
        );
        for resource in &resources {
            self.cache
                .update_object(&resource.object.id, resource.clone());
        }
        let project: generic_resource::Resource =
            // First entry is always the archived project 
            tonic_internal!(resources[0].clone().try_into(), "Project conversion error");
```

File: src/grpc/collections.rs [hunk 7]
```diff
@@ -236,35 +286,55 @@ impl CollectionService for CollectionServiceImpl {
...
+            self.database_handler.snapshot(request).await,
+            "Internal database error."
+        );
+        for resource in &resources {
+            self.cache
+                .update_object(&resource.object.id, resource.clone());
+        }
+        let collection: generic_resource::Resource =
+        // First entry is always the snapshot collection
+            tonic_internal!(resources[0].clone().try_into(), "Dataset conversion error");
```
Comment (das-Abroxas) [inline src/grpc/collections.rs:333]: Copy & paste only partially adapted.

```suggestion
        let collection_id = tonic_invalid!(request.get_id(), "Invalid collection id.");
        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::ADMIN, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        // This contains all resources the snapshot includes below the collection
        let resources = tonic_internal!(
            self.database_handler.snapshot(request).await,
            "Internal database error."
        );
        for resource in &resources {
            self.cache
                .update_object(&resource.object.id, resource.clone());
        }
        let collection: generic_resource::Resource =
        // First entry is always the snapshot collection
            tonic_internal!(resources[0].clone().try_into(), "Collection conversion error");
```

**Review by das-Abroxas — CHANGES_REQUESTED** [review]
Some copy & paste things that should be smoothed out before merge.

**Review by St4NNi — APPROVED** [review]
Nice, additions for now LGTM. This will surely be modified in further developments.

**Review by das-Abroxas — APPROVED** [review]

**Pull request merged**